### PR TITLE
MGMT-1695 Support HTTPS transport

### PR DESCRIPTION
**Pull request by empovit**
This PR adds control over the default HTTPS transport by adding `--insecure` option (skip certificate validation) and `--cacert` option (provide a custom root CA certificate as a PEM file).

### CODE REVIEW

File: src/session/inventory_session.go [hunk 1]
```diff
@@ -35,21 +40,70 @@ func (i *InventorySession) Client() *client.AssistedInstall {
...
+	certs, err := readCACertificate()
+	if err != nil {
+		return nil, err
+	}
+
+	clientConfig.Transport = requestid.Transport(&http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
```
Comment (mhrivnak) [inline src/session/inventory_session.go:60]: Any particular reason to change from the default of 15s?

File: README.md [hunk 2]
```diff
@@ -48,6 +50,14 @@ The subsystem tests uses docker-compose to run the agent and wiremock simulating
 
 To perform the subsystem test run: `skipper make subsystem`
 
+> **Note:** If many subsystem tests fail unexpectedly on Fedora/RHEL, it is likely a firewall blocks
```
Comment (ori-amizur) [inline README.md:53]: Might be irrelevant now after last changes https://github.com/openshift/assisted-installer-agent/pull/6

File: src/config/agent_config.go [hunk 3]
```diff
@@ -42,6 +44,8 @@ func ProcessArgs() {
 	flag.StringVar(&ret.InventoryImage, "inventory-image", "quay.io/ocpmetal/inventory:latest", "The image of inventory")
 	flag.BoolVar(&ret.JournalLogging, "with-journal-logging", true, "Use journal logging")
 	flag.BoolVar(&ret.TextLogging, "with-text-logging", false, "Output log to file")
+	flag.StringVar(&ret.CACertificate, "cacert", "", "Path to custom CA certificate in PEM format")
+	flag.BoolVar(&ret.InsecureConnection, "insecure", false, "Do not validate TLS certificate")
```
Comment (mhrivnak) [inline src/config/agent_config.go:48]: I think it's worth a log statement somewhere when this is set to `true`.

File: src/session/inventory_session.go [hunk 4]
```diff
@@ -35,21 +40,70 @@ func (i *InventorySession) Client() *client.AssistedInstall {
...
 	clientConfig := client.Config{}
-	clientConfig.URL, _ = url.Parse(createUrl())
-	clientConfig.Transport = requestid.Transport(http.DefaultTransport)
+	var err error
+	clientConfig.URL, err = url.ParseRequestURI(createUrl())
+	if err != nil {
+		return nil, err
+	}
+
+	certs, err := readCACertificate()
```
Comment (ori-amizur) [inline src/session/inventory_session.go:51]: Why are attempting to read ca certificate if we work in insecure mode?

File: src/session/inventory_session.go [hunk 5]
```diff
@@ -35,21 +40,70 @@ func (i *InventorySession) Client() *client.AssistedInstall {
...
+	})
+
 	clientConfig.AuthInfo = auth.AgentAuthHeaderWriter(config.GlobalAgentConfig.PullSecretToken)
 	bmInventory := client.New(clientConfig)
-	return bmInventory
+	return bmInventory, nil
 }
 
-func New() *InventorySession {
+func readCACertificate() (*x509.CertPool, error) {
```
Comment (ori-amizur) [inline src/session/inventory_session.go:78]: Do we assume that the CACertificate is root CA?

File: README.md (hunk 2 above)
Comment (empovit) [reply README.md:53]: Removing, thanks!

File: src/config/agent_config.go [hunk 6]
```diff
@@ -22,6 +22,8 @@ var GlobalAgentConfig struct {
 	TextLogging        bool
 	AgentVersion       string
 	PullSecretToken    string
+	InsecureConnection bool
+	CACertificate      string
```
[the remark this answers is not in the document]
Comment (empovit) [reply src/config/agent_config.go:26]: Good point, will change

File: src/session/inventory_session.go (hunk 1 above)
Comment (empovit) [reply src/session/inventory_session.go:60]: Hmmm, the default is 30 https://github.com/golang/go/blob/release-branch.go1.13/src/net/http/transport.go#L45. I copied DefaultTransport as is, just removed DualStack, which is deprecated.

File: src/config/agent_config.go (hunk 3 above)
Comment (empovit) [reply src/config/agent_config.go:48]: Will do

File: src/session/inventory_session.go (hunk 4 above)
Comment (empovit) [reply src/session/inventory_session.go:51]: Right, will be fixed.

File: src/session/inventory_session.go (hunk 5 above)
Comment (empovit) [reply src/session/inventory_session.go:78]: Theoretically any server cert can be passed, but it CA makes more sense.

File: src/session/inventory_session.go (hunk 1 above)
Comment (mhrivnak) [reply src/session/inventory_session.go:60]: The [docs here](https://golang.org/pkg/net/#Dialer) say 15s is the default for the `Dialer`. It's interesting that `DefaultTransport` changes that default. Not a big deal, and I'm sure either is fine.

File: src/session/inventory_session.go (hunk 5 above)
Comment (empovit) [reply src/session/inventory_session.go:78]: The corresponding field of `TLSClientConfig` is called `RootCAs`.

File: src/session/inventory_session.go (hunk 1 above)
Comment (empovit) [reply src/session/inventory_session.go:60]: Interesting. In this particular case I will probably keep the behavior that was already tested (i.e. http.DefaultTransport).

**Pull request merged**